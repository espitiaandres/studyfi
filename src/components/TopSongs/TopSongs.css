.songsTableTitle {
    color: white;
    font-size: 36px;
}

.songsTableTimeRanges {
    align-items: center;
    display: flex;
    justify-content: center;
    padding-bottom: 25px;
}

.songsTableTimeRangeButton {
    background: #1ED760;
    border: none;
    border-radius: 5px;
    display: inline-block;
    margin: 0px 20px;    
    padding: 10px;
    text-decoration: none;
    min-width: 160px;
}

.songsTableTimeRangeButton:hover {
    cursor: pointer;
    filter: brightness(70%);
}

.songsTable {
    color: white;
    padding: 0px 80px;
}

.songsTablePagination {
    background: #1ED760;
    border: none;
    border-radius: 1px;
    display: inline-block;
    font-size: 16px;
    margin: 0px 4px;    
    text-decoration: none;
    width: 30px;
}

.songsTablePagination:hover {
    cursor: pointer;
    filter: brightness(70%);
}  
.columnDescriptions {
    padding-bottom: 20px;
}

.tooltips {
    max-width: 400px;
}

.tooltip {
    background: #1ED760;
    border: none;
    border-radius: 1px;
    display: inline-block;
    font-size: 16px;
    margin: 0px 4px;  
    padding: 2px 10px;  
    text-decoration: none;
}

.seasonStyling {
    background: #F37D0F;
}

.seasonStylingAlt {
    background: #FFE01B;
}
