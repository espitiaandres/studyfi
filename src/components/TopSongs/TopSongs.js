//
//  TopSongs.js
//  react-spotify-player
//
//  Created by Andres Espitia.
//  Copyright Â© 2020 Andres Espitia. All rights reserved.
//

import React, { useState, useEffect } from 'react';
import axios from 'axios';
import BootstrapTable from 'react-bootstrap-table-next';
import hash from '../../utils/hash';
import './TopSongs.css';

const TopSongs = () => {
    const token = hash.access_token;
    let filteredTopTracksData = [];
    const [filteredData, setFilteredData] = useState([]);


    useEffect(() => {
        let audioFeaturesTracksIDs = "";

        axios({
            method: 'get',
            url: 'https://api.spotify.com/v1/me/top/tracks',
            headers: {
            'Authorization': `Bearer ${token}`
            },
            params: {
                limit: 50,
                time_range: 'long_term'
            }
        }).then(({ data }) => {
            data.items.map((song) => {
                let allArtists = "";
                const topTrackFiltered = {};        
                topTrackFiltered.name = song.name;
                topTrackFiltered.album = song.album.name;
                topTrackFiltered.albumCover = song.album.images[2].url;
                topTrackFiltered.releaseDate = song.album.release_date;
                topTrackFiltered.id = song.id;
                audioFeaturesTracksIDs += song.id + ",";
                song.artists.map((artist) => {
                    allArtists += `${artist.name}, `;
                })
                topTrackFiltered.artists = allArtists.slice(0, -2);
                filteredTopTracksData.push(topTrackFiltered);
            });

            audioFeaturesTracksIDs = audioFeaturesTracksIDs.slice(0, -1);

            // following link explains what each category indicates.
            // https://developer.spotify.com/documentation/web-api/reference/tracks/get-several-audio-features/

            axios({
                method: 'get',
                url: 'https://api.spotify.com/v1/audio-features/',
                headers: {
                'Authorization': `Bearer ${token}`
                },
                params: {
                    ids: audioFeaturesTracksIDs
                }
            }).then(({ data: { audio_features } }) => {
                for (let i = 0; i < filteredTopTracksData.length; i++) {
                    filteredTopTracksData[i].acousticness = audio_features[i].acousticness;
                    filteredTopTracksData[i].danceability = audio_features[i].danceability;
                    filteredTopTracksData[i].energy = audio_features[i].energy;
                    filteredTopTracksData[i].keySignature = audio_features[i].key;
                    filteredTopTracksData[i].instrumentalness = audio_features[i].instrumentalness;
                    filteredTopTracksData[i].valence = audio_features[i].valence;
                    filteredTopTracksData[i].key = i + 1;
                }

                setFilteredData(filteredTopTracksData);
            })
        })
    }, []);

    function formatter(cell, row) {     
        return cell.toString().includes("http") ? 
        <span>
            <img src={cell}/>
        </span> : 
        <span>
            {cell}
        </span>;
    }

    const columns = [
        {
            text: 'Name',
            dataField: 'name',
            formatter
        },
        {
            text: 'Album',
            dataField: 'album',
            formatter
        },
        {
            text: 'Album Cover',
            dataField: 'albumCover',
            formatter
        },
        {
            text: 'Release Date',
            dataField: 'releaseDate',
            formatter
        },
        {
            text: 'Artists',
            dataField: 'artists',
            formatter
        },
        {
            text: 'Acousticness',
            dataField: 'acousticness',
            formatter
        },
        {
            text: 'Danceability',
            dataField: 'danceability',
            formatter
        },
        {
            text: 'Energy',
            dataField: 'energy',
            formatter
        },
        {
            text: 'KeySignature',
            dataField: 'keySignature',
            formatter
        },
        {
            text: 'Instrumentalness',
            dataField: 'instrumentalness',
            formatter
        },
        {
            text: 'Valence',
            dataField: 'valence',
            formatter
        }
    ]

    return (
        <div>
            <h1>your most listened to artists and tracks</h1>
            <div>
                <BootstrapTable
                    keyField="id"
                    data={filteredData}
                    columns={columns}
                />
            </div>
        </div>
    )
}

export default TopSongs
