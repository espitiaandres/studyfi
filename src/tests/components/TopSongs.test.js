//
//  TopSongs.test.js
//  react-spotify-player
//
//  Created by Andres Espitia.
//  Copyright Â© 2020 Andres Espitia. All rights reserved.
//

import React from 'react';
import { shallow } from 'enzyme';
import TopSongs from '../../components/TopSongs/TopSongs';

const topSongsPropsTest = {
    season: true,
    token: "mytoken"
}

test('should render TopSongs component correctly', () => {
    const wrapper = shallow(<TopSongs />);
    expect(wrapper).toMatchSnapshot();
});

test('should render TopSongs component correctly', () => {
    const wrapper = shallow(
        <TopSongs 
            season={topSongsPropsTest.season}
            token={topSongsPropsTest.token}
        />
    );
    expect(wrapper).toMatchSnapshot();
});