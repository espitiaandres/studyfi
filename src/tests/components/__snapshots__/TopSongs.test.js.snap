// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TopSongs component correctly 1`] = `ShallowWrapper {}`;

exports[`should render TopSongs component correctly 2`] = `ShallowWrapper {}`;
